src/main/java/com/google/devtools/build/lib/bazel/repository/cache/RepoContentsCache.java

```diff
@@ -128,4 +183,81 @@ private static String getNextCounterInDir(Path entryDir) throws IOException {
       return counter;
     }
   }
+
+  public void acquireSharedLock() throws IOException {
+    Preconditions.checkState(path != null);
+    Preconditions.checkState(sharedLock == null, "this process already has the shared lock");
+    sharedLock = FileSystemLock.get(path.getRelative(LOCK_PATH), LockMode.SHARED);
```

Comment:
Consider adding a profile trace here, so that time spent waiting for the lock is apparent.

Reply:
addressed during import